### PR TITLE
Automate generation of javadocs and create PR with changes

### LINKED ISSUE #12389: [Documentation Request] Java API docs need updating
Java docs are out of date. SessionOptions is missing addConfigEntry. 

---
#### Document Details
* Title: API Docs
* Page: [https://onnxruntime.ai/docs/api/](https://onnxruntime.ai/docs/api/)
* Page Source: index.md

### PR DESCRIPTION
**Pull request by natke**
Fixed #12389 

Staged PR here: https://github.com/natke/onnxruntime/pull/29

### CODE REVIEW

**Comment by natke** [comment]
@Craigacp fyi

File: .github/workflows/publish-java-apidocs.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  push:
+    branches:
+      - master
+  workflow_dispatch:
+
+jobs:
+  publish:
+    name: Generate Java docs
+    runs-on: ubuntu-latest
+    steps:
```
Comment (skottmckay) [inline .github/workflows/publish-java-apidocs.yml:12]: Is this specifying a docker container which has the ORT source and can be used to build the javadoc? If so where is `ubuntu-latest` defined?

File: .github/workflows/publish-java-apidocs.yml [hunk 2]
```diff
@@ -0,0 +1,48 @@
+name: Update Java API Docs
+on:
+  push:
+    branches:
+      - master
+  workflow_dispatch:
```
Comment (skottmckay) [inline .github/workflows/publish-java-apidocs.yml:6]: Does this need to change to 'main'?

**Comment by skottmckay** [comment]
Is there doco somewhere on how/when this workflow runs? Not sure if it runs automatically or would it be part of the release process?

**Comment by Craigacp** [comment]
> @Craigacp fyi

The generated docs look good. There are a few missing docstrings as it's not pulling them in automatically from overridden methods in the JDK (e.g. places which override `toString` or `close` usually pull in the Javadoc from those methods on `java.lang.Object` and `java.lang.AutoCloseable` but it's not doing that for some reason), and it's missing a `package-info.java` in the `ai.onnxruntime.providers` package. I can add those in another PR once the outstanding JNI PRs are merged in (as there will be conflicts due to some new Java methods in the JNI refactor).

I'm not sure how to feed this into the Gradle javadoc plugin, but it would be nice to have an equivalent `overview.html` snippet like we have in the Tribuo docs to put a useful description on the Javadoc landing page (e.g. https://github.com/oracle/tribuo/blob/main/Core/src/main/javadoc/overview.html -> the header text above the package list here - https://tribuo.org/learn/4.2/javadoc/).

File: .github/workflows/publish-java-apidocs.yml (hunk 1 above)
Comment (natke) [reply .github/workflows/publish-java-apidocs.yml:12]: ubuntu-latest is the container that the GitHUb actiions runs in. The ort source is checked out separately

File: .github/workflows/publish-java-apidocs.yml (hunk 2 above)
Comment (natke) [reply .github/workflows/publish-java-apidocs.yml:6]: OOh yes it does, well spotted

**Comment by natke** [comment]
> Is there doco somewhere on how/when this workflow runs? Not sure if it runs automatically or would it be part of the release process?

It runs on every push to main/master, so changes are picked up straight away

**Comment by skottmckay** [comment]
> > Is there doco somewhere on how/when this workflow runs? Not sure if it runs automatically or would it be part of the release process?
> 
> It runs on every push to main/master, so changes are picked up straight away

With this setup do we have any way to see what the API was for a specific release? Or will the doco only be for the very latest code which may not match what the user has? Looks like that's what we do for the other APIs as well though.

**Comment by natke** [comment]
> > > Is there doco somewhere on how/when this workflow runs? Not sure if it runs automatically or would it be part of the release process?
> > 
> > 
> > It runs on every push to main/master, so changes are picked up straight away
> 
> With this setup do we have any way to see what the API was for a specific release? Or will the doco only be for the very latest code which may not match what the user has? Looks like that's what we do for the other APIs as well though.

Yes that is what we decided. Given that our APIs are backwards compatible and do not change much for each release, we note in the docs where specific methods and parameters are introduced in which version. There are javadoc annotations for this.

**Comment by natke** [comment]
> > @Craigacp fyi
> 
> The generated docs look good. There are a few missing docstrings as it's not pulling them in automatically from overridden methods in the JDK (e.g. places which override `toString` or `close` usually pull in the Javadoc from those methods on `java.lang.Object` and `java.lang.AutoCloseable` but it's not doing that for some reason), and it's missing a `package-info.java` in the `ai.onnxruntime.providers` package. I can add those in another PR once the outstanding JNI PRs are merged in (as there will be conflicts due to some new Java methods in the JNI refactor).
> 
> I'm not sure how to feed this into the Gradle javadoc plugin, but it would be nice to have an equivalent `overview.html` snippet like we have in the Tribuo docs to put a useful description on the Javadoc landing page (e.g. https://github.com/oracle/tribuo/blob/main/Core/src/main/javadoc/overview.html -> the header text above the package list here - https://tribuo.org/learn/4.2/javadoc/).

I *think* (have not confirrmed) that if we added the overview.html into a javadoc dir in our java/src/main dir that  this would build this into the generated docs.

**Comment by Craigacp** [comment]
Ok. I guess I should do a pass to add the `@since` tags and other things to the Javadoc. I can try adding the overview to see if it works.

**Comment by natke** [comment]
> Ok. I guess I should do a pass to add the `@since` tags and other things to the Javadoc. I can try adding the overview to see if it works.

Shall we get this PR merged, and make the updates in a subsequent one then?

**Comment by Craigacp** [comment]
Yes, I don't have time to do that this week.

**Comment by natke** [comment]
@skottmckay Are you ok to approve the review? Or do you want me to add more comments to the action?

**Review by skottmckay — APPROVED** [review]

**Pull request merged**